### DISCUSSION TITLE
LVGL Example Arduino Config files

### DISCUSSION THREAD

**Discussion by VollWombat** [Westcott1/Waveshare-ESP32-S3-Touch-LCD-4.3-and-Arduino#1] [General]
Hey, I found this project via Google and thought I could add something to it.
So, I tried to get the provided demos (The "Sample demos" link from [here](https://www.waveshare.com/wiki/ESP32-S3-Touch-LCD-4.3#Demo)) to run and had some issues with it. 

So for documentation, those are the steps I did to get the sample "lvgl_porting" from the package above to run (mostly based on the Waveshare Wiki I linked above):

Adding the espressif arduino esp32 repo to Arduino IDE:
File -> Preferences -> Additional Boards Managers URLs -> Add "https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json")

Installing the required libraries:
- ESP32_Display_Panel by espressif [version 0.0.2 as of now] 
- ESP32_IO_Expander by lzw655 [version 0.0.2 as of now] 
- lvgl by kisvegabor [version 8.3.11 as of now]

Afterwards, you should be able to select the Board via Tools -> Board <Currentboardname> -> esp32 -> ESP32S3 Dev Module.

Now that was  all no really big deal, but I found no info on how to configure the ESP32_Display_Panel and lvgl library. They have default config files, but the samples didn't run with them. So I wrote back and forth with the support about this issue (and some others I had initially), where after a few weeks I got the sample and the libraries they used, including their configuration files.

[ESP_Panel_Conf.txt](https://github.com/Westcott1/Waveshare-ESP32-S3-Touch-LCD-4.3-and-Arduino/files/13931682/ESP_Panel_Conf.txt)
[lv_conf.txt](https://github.com/Westcott1/Waveshare-ESP32-S3-Touch-LCD-4.3-and-Arduino/files/13931683/lv_conf.txt)
Those files need to be renamed (remove the .txt at the end and replace it with .h) and put them into My Documents\Arduino\libraries\ directory (directly there, not in the individual folders of the packages itself). 
If you don't want to use these files, read further below where I listed all the changes to the default config file.

Also, for those sample, at least the lvgl demos are needed. In order to use them, you need to copy the folder <My documents>\Arduino\libraries\lvgl\demos to <My documents>\Arduino\libraries\lvgl\**src\**demos.
Better safe than sorry, you can also copy the examples-folder in from the same parent directory to the src-folder.

If you don't want to use my files, you can do the following:
Copy **and rename** the file "Documents\Arduino\libraries\lvgl\lv_conf_template.h" to Documents\Arduino\libraries\lv_conf.h
Edit the file and replace the following lines:
`#if 0 /*Set it to "1" to enable content*/` becomes `#if 1 /*Set it to "1" to enable content*/`
for the following lines, just search for the variable name (the one that comes after #define) and replace the value after it with the value mentioned here:
```
#define LV_MEM_CUSTOM 1
#define LV_MEMCPY_MEMSET_STD 1
#define LV_TICK_CUSTOM 1
#define LV_ATTRIBUTE_FAST_MEM IRAM_ATTR   // Get higher performance but take up more SRAM
#define LV_FONT_MONTSERRAT_N    1  // Enable all the internally used fonts (`N` should be replaced by the font size)
#define LV_USE_DEMO_WIDGETS 1
```

Next for ESP32 Display Panel, copy **and rename** the file 'Documents\Arduino\libraries\ESP32_Display_Panel\ESP_Panel_Conf_Template.h' to 'Documents\Arduino\libraries\ESP_Panel_Conf.h'
and make the changes just like in lv_conf.h with the values used below:
```
#define ESP_PANEL_USE_SUPPORTED_BOARD   (0)
#define ESP_PANEL_USE_LCD           (1)
#define ESP_PANEL_LCD_NAME          ST7262
#define ESP_PANEL_LCD_H_RES         (800)
#define ESP_PANEL_LCD_V_RES         (480)
#define ESP_PANEL_LCD_BUS_SKIP_INIT_HOST        (1)
#define ESP_PANEL_LCD_BUS_TYPE      (3)
#define ESP_PANEL_LCD_RGB_CLK_HZ            (14 * 1000 * 1000)
#define ESP_PANEL_LCD_RGB_IO_DE             (5)
#define ESP_PANEL_LCD_RGB_IO_PCLK           (7)
#define ESP_PANEL_LCD_RGB_IO_DATA0          (14)
#define ESP_PANEL_LCD_RGB_IO_DATA1          (38)
#define ESP_PANEL_LCD_RGB_IO_DATA2          (18)
#define ESP_PANEL_LCD_RGB_IO_DATA3          (17)
#define ESP_PANEL_LCD_RGB_IO_DATA4          (10)
#define ESP_PANEL_LCD_RGB_IO_DATA5          (39)
#define ESP_PANEL_LCD_RGB_IO_DATA6          (0)
#define ESP_PANEL_LCD_RGB_IO_DATA7          (45)
#define ESP_PANEL_LCD_RGB_IO_DATA8          (48)
#define ESP_PANEL_LCD_RGB_IO_DATA9          (47)
#define ESP_PANEL_LCD_RGB_IO_DATA10         (21)
#define ESP_PANEL_LCD_RGB_IO_DATA11         (1)
#define ESP_PANEL_LCD_RGB_IO_DATA12         (2)
#define ESP_PANEL_LCD_RGB_IO_DATA13         (42)
#define ESP_PANEL_LCD_RGB_IO_DATA14         (41)
#define ESP_PANEL_LCD_RGB_IO_DATA15         (40)
#define ESP_PANEL_USE_LCD_TOUCH     (1)
#define ESP_PANEL_LCD_TOUCH_NAME            GT911
#define ESP_PANEL_LCD_TOUCH_I2C_IO_SCL      (9)
```

And I think that should be it. Afterwards the lvgl_porting.ino demo should run. Hope that helps someone!

**Comment by Westcott1** [comment]
WOW, thanks VollWombat!

I would have replied sooner, but my son and his wife just brought round their firstborn, our day-old grandson.
I'm a granddad!

Meanwhile, back in Git-land:
You have taken a somewhat different route to me.
Yes, I have got LVGL running successfully.
However, I never found an Arduino demo from Waveshare, despite asking their support for one.
Their demo is an Espressiv 'c' code, but it contain some useful bits.
Also I cannot get the ESP_IO_Expander example to work.
The special version for this board does include support for the CH422G expander chip in ESP_IOExpander_Library.h, but I can't see how to include it.

Thanks again,
Hamish

**Comment by VollWombat** [comment]
Hi Westcott and congratulations on your grandchild! :-)

You're totally right, I had the same issue and while the support couldn't give me a clear answer to this, I just noticed that the ESP_IO_Expander library they send me had indeed 2 files more, related to exactly that chip.

[chip_CH422G.zip](https://github.com/Westcott1/Waveshare-ESP32-S3-Touch-LCD-4.3-and-Arduino/files/13932259/chip_CH422G.zip)

Put those two files into Documents\Arduino\libraries\ESP32_IO_Expander\src\chip\.
Then, modify the file "Documents\Arduino\libraries\ESP32_IO_Expander\src\ESP_IOExpander_Library.h" and put the following line before the last #endif:
`#include "chip/CH422G.h"`

Then you should be good to go.

Regarding the arduino demo, I think they added it later (maybe they made it for me when I contacted them and added it afterwards 😅). Its now included in the "Sample demo" download on the [ESP32-S3-Touch-LCD-4.3 Wiki page](https://www.waveshare.com/wiki/ESP32-S3-Touch-LCD-4.3#Demo)

**Reply by Westcott1** [reply]
Hi VollWombat.
Thanks!
Regarding ES32_IO_Expander, yes, I had all of those files and #include in place.
What I can't see is how to get it to include and use ESP_IOExpander_Library.h
I'd also expect a constructor or #define to say which chip to use.

Yes, I see that ESP32-S3-Touch-LCD-4.3_Code.zip now includes an Arduino tree.

UPDATE
The lvgl_Porting example has #include <ESP_IOExpander_Library.h> and constructor.

**Reply by VollWombat** [reply]
Yeah I was just about to write that :) The lvgl_porting example should run without any further modification.

**Comment by bpRF-ENG** [comment]
Congratulations on being a grandfather!

I don't know if you can help me, when I run the INO file, I receive this error message.

Compilation error: 'Arduino_ESP32RGBPanel' does not name a type; did you mean 'Arduino_ESP32S2PAR8Q'?

Thank you,

**Reply by Westcott1** [reply]
Thanks!
Are you using Arduino_GFX_Library.h rather than ESP_Panel_Library.h ?
I think that your compiler error is caused by Arduino_GFX_Library incompatibility with Arduino v3.

**Reply by bpRF-ENG** [reply]
I will check that out. 

Thank you for the quick response.

**Comment by sjgcornish** [comment]
I finally managed to get to get this working with Arduino IDE. I followed the steps in Vaveshare wiki. It was just a matter of getting the right files in the right place. Note the porting example contains the correct ESP_Panel_Conf.h and lv_conf.h. 
To get the lv_porting to work did require some changes...
Copy the Waveshare demos folder tot he lvgl SRC folder ( as mention in the lv_porting.h file)
Create a folder lvgl_port_v8 in library and move the lvgl_port_v8.h and lvgl_port_v8.cpp files to it.
in lvgl_porting.ino change #include "lvgl_port_v8.h" to #include <lvgl_port_v8.h>
For some reason the .ino was reporting the definitions in lvgl_port_v8.h where not being picked up by the compiler and the above was my workaround.
